service/src/DCGMIOGroup.cpp

```diff
@@ -0,0 +1,541 @@
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": domain_idx out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        double result = NAN;
+        m_dcgm_device_pool.update(domain_idx);
+
+        auto it = m_signal_available.find(signal_name);
+        if (it != m_signal_available.end()) {
+            result = it->second.m_devpool_func(domain_idx);
+        }
+        else {
+    #ifdef GEOPM_DEBUG
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": Handling not defined for " +
+                            signal_name, GEOPM_ERROR_LOGIC, __FILE__, __LINE__);
+    #endif
+        }
+        return result;
+    }
+
+    // Write to the control immediately, bypassing write_batch()
+    void DCGMIOGroup::write_control(const std::string &control_name, int domain_type, int domain_idx, double setting)
+    {
+        if (!is_valid_control(control_name)) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": " + control_name +
+                            " not valid for DCGMIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_type != control_domain_type(control_name)) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": " + control_name + ": domain_type must be " +
+                            std::to_string(control_domain_type(control_name)),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_idx < 0 || domain_idx >= m_platform_topo.num_domain(GEOPM_DOMAIN_BOARD)) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": domain_idx out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        if (control_name == "DCGM::FIELD_UPDATE_RATE") {
+            m_dcgm_device_pool.update_rate(setting*1e6);
+        }
+        else if (control_name == "DCGM::MAX_STORAGE_TIME") {
+            m_dcgm_device_pool.max_storage_time(setting);
+        }
+        else if (control_name == "DCGM::MAX_SAMPLES") {
+            m_dcgm_device_pool.max_samples(setting);
+        }
+        else {
+    #ifdef GEOPM_DEBUG
+                throw Exception("DCGMIOGroup::" + std::string(__func__) + "Handling not defined for "
+                                + control_name, GEOPM_ERROR_LOGIC, __FILE__, __LINE__);
+    #endif
+        }
+    }
+
+    // Implemented to allow an IOGroup to save platform settings before starting
+    // to adjust them
+    void DCGMIOGroup::save_control(void)
+    {
+    }
+
+    // Implemented to allow an IOGroup to restore previously saved
+    // platform settings
+    void DCGMIOGroup::restore_control(void)
+    {
+        m_dcgm_device_pool.update_rate(1*1e6); //1 second
+        m_dcgm_device_pool.max_samples(0); //0 = no limit
+        m_dcgm_device_pool.max_storage_time(10); //10 seconds.
```

Comment:
Are these published default values?  What is the source for these values?

Reply:
The update rate is the default value specified by the DCGM user guide [here](https://docs.nvidia.com/datacenter/dcgm/latest/dcgm-user-guide/feature-overview.html#profiling-metrics)

The maximum samples (DCGM variable maxKeepSamples) and maximum storage time (DCGM variable maxKeepAge) don't appear to have a published default value, so a 'sensible' value was chosen from our testing.  The dcgmWatchFields interface using these values is [here](https://docs.nvidia.com/datacenter/dcgm/latest/dcgm-api/group__DCGMAPI__FI.html#group__DCGMAPI__FI_1gd04fa06405af6165fd85310dee9eae2a)

Reply:
This has been further tweaked to use the 'example' values provided in the DCGM codebase as no other default value could be found.